BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Authentication/ReflectionCredentialEnvelope.cs

```diff
@@ -0,0 +1,84 @@
+                // Invoking this constructor: Azure.Core.TokenRequestContext(String[], String, String).
+                var tokenRequestContextType = Type.GetType("Azure.Core.TokenRequestContext, Azure.Core");
+                var paramArray = new object[] { AuthConstants.GetScopes(), null, null };
+                this.tokenRequestContext = Activator.CreateInstance(tokenRequestContextType, args: paramArray);
+
+                // (https://docs.microsoft.com/en-us/dotnet/api/azure.core.tokencredential.gettokenasync?view=azure-dotnet).
+                // Getting a handle for this method: Azure.Core.TokenCredential.GetTokenAsync().
+                this.getTokenAsyncMethod = this.Credential.GetType().GetMethod("GetTokenAsync");
+                this.getTokenMethod = this.Credential.GetType().GetMethod("GetToken");
+
+                // this.accessTokenType = Type.GetType("Azure.Core.AccessToken, Azure.Core");
+            }
+            else
+            {
+                throw new ArgumentException($"The provided {nameof(tokenCredential)} must inherit Azure.Core.TokenCredential", nameof(tokenCredential));
+            }
+        }
+
+        public object Credential { get; private set; }
+
+        public string GetToken(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var accessToken = this.getTokenMethod.Invoke(this.Credential, new object[] { this.tokenRequestContext, cancellationToken });
+            var tokenProperty = accessToken.GetType().GetProperty("Token");
+            return (string)tokenProperty.GetValue(accessToken);
+        }
+
+        public async Task<string> GetTokenAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await Task.Run(() => this.GetToken(cancellationToken));
+
+            // TODO: THIS DOESN'T WORK. CAN CIRCLE BACK AND INVESTIGATE THIS LATER.
+            // 'Unable to cast object of type 'System.Threading.Tasks.ValueTask`1[Azure.Core.AccessToken]' to type 'System.Threading.Tasks.Task'.'
+            // (https://stackoverflow.com/questions/39674988/how-to-call-a-generic-async-method-using-reflection/39679855).
+            //var task = (Task)this.getTokenAsyncMethod.Invoke(this.Credential, new object[] { this.tokenRequestContext, cancellationToken });
+            //await task.ConfigureAwait(false);
+            //var resultProperty = task.GetType().GetProperty("Result");
+            //var accessToken = resultProperty.GetValue(task);
+            //var tokenProperty = accessToken.GetType().GetProperty("Token");
+            //return (string)tokenProperty.GetValue(accessToken);
+        }
+
+        /// <summary>
+        /// Checks if the input inherits <see cref="Azure.Core.TokenCredential"/>
+        /// </summary>
+        private bool IsTokenCredential(Type inputType) => inputType.IsSubclassOf(Type.GetType("Azure.Core.TokenCredential, Azure.Core"));
```

Comment:
`Type.GetType("Azure.Core.TokenCredential, Azure.Core")` needs to be precalcualted/cached so we don't end up calling reflection each time.

Reply:
This will only be called once (from the constructor) so I think we're safe to not cache this one.

Reply:
An earlier version of this method was more complicated before I found the one-line statement (verified with unit tests). 
I could move this into the constructor now to eliminate the confusion. :)

Reply:
If the constructor is supposed to be called once (e.g. singleton) it's fine.